File: app/src/test/java/de/cineaste/android/util/MultiListTest.java
package de.cineaste.android.util;

import org.junit.Before;
import org.junit.Test;

import java.util.ArrayList;
import java.util.List;

import de.cineaste.android.entity.movie.MovieDto;

import static org.junit.Assert.*;

public class MultiListTest {
    private final List<MovieDto> initialDtos = new ArrayList<>();
    private MultiList list;

    @Before
    public void init() {
        list = new MultiList();

        for (int i = 0; i < 5; i++) {
            initialDtos.add(new MovieDto(i, "path_" + i, "title " + i));
        }
    }

    @Test
    public void shouldReturnOneEntry() {
        list.add(initialDtos.get(0));

        assertEquals(1, list.getMovieList().size());
    }

    @Test
    public void shouldReturn5EntriesWithCounter1() {
        list.addAll(initialDtos);

        assertEquals(5, list.getMovieList().size());
        for (MultiList.MultiListEntry multiListEntry : list.getMovieList()) {
            assertEquals(1, multiListEntry.getCounter());
        }
    }

    @Test
    public void shouldReturnEntriesOrderedByCounter() {
        list.add(initialDtos.get(0));

        list.add(initialDtos.get(1));

        list.add(initialDtos.get(2));
        list.add(initialDtos.get(2));
        list.add(initialDtos.get(2));
        list.add(initialDtos.get(2));

        list.add(initialDtos.get(3));
        list.add(initialDtos.get(3));

        list.add(initialDtos.get(4));
        list.add(initialDtos.get(4));
        list.add(initialDtos.get(4));

        assertEquals(5, list.getMovieList().size());

        MultiList.MultiListEntry first = list.getMovieList().get(0);
        MultiList.MultiListEntry second = list.getMovieList().get(1);
        MultiList.MultiListEntry third = list.getMovieList().get(2);
        MultiList.MultiListEntry fourth = list.getMovieList().get(3);
        MultiList.MultiListEntry fifth = list.getMovieList().get(4);

        assertEquals(initialDtos.get(2), first.getMovieDto());
        assertEquals(4, first.getCounter());

        assertEquals(initialDtos.get(4), second.getMovieDto());
        assertEquals(3, second.getCounter());

        assertEquals(initialDtos.get(3), third.getMovieDto());
        assertEquals(2, third.getCounter());

        assertEquals(fourth.getCounter(), 1);
        assertEquals(fifth.getCounter(), 1);
    }

}

File: app/src/test/kotlin/de/cineaste/android/util/MultiListTest.kt
package de.cineaste.android.util

import org.junit.Before
import org.junit.Test

import java.util.ArrayList

import de.cineaste.android.entity.movie.MovieDto

class MultiListTest {
    private val initialDtos = ArrayList<MovieDto>()
    private lateinit var list: MultiList

    @Before
    fun init() {
        list = MultiList()

        val title = "ABCDE"

        for (i in 0..4) {
            initialDtos.add(MovieDto(i.toLong(), "path_$i", "${title[i]} title $i"))
        }
    }

    @Test
    fun shouldReturnOneEntry() {
        list.add(initialDtos[0])

        assertEquals(1, list.getSortedList().size.toLong())
    }

    @Test
    fun shouldReturn5EntriesWithCounter1() {
        list.addAll(initialDtos)

        assertEquals(5, list.getSortedList().size.toLong())

        list.getSortedList().forEachIndexed { index, multiListEntry ->
            assertEquals(1, multiListEntry.counter)
            assertEquals(initialDtos[index], multiListEntry.movieDto)
        }
    }

    @Test
    fun shouldReturnEntriesOrderedByCounterAndName() {
        list.add(initialDtos[0])

        list.add(initialDtos[1])

        list.add(initialDtos[2])
        list.add(initialDtos[2])
        list.add(initialDtos[2])
        list.add(initialDtos[2])

        list.add(initialDtos[3])
        list.add(initialDtos[3])

        list.add(initialDtos[4])
        list.add(initialDtos[4])
        list.add(initialDtos[4])

        assertEquals(5, list.getSortedList().size.toLong())

        val first = list.getSortedList()[0]
        val second = list.getSortedList()[1]
        val third = list.getSortedList()[2]
        val fourth = list.getSortedList()[3]
        val fifth = list.getSortedList()[4]

        assertEquals(initialDtos[2], first.movieDto)
        assertEquals(4, first.counter.toLong())

        assertEquals(initialDtos[4], second.movieDto)
        assertEquals(3, second.counter.toLong())

        assertEquals(initialDtos[3], third.movieDto)
        assertEquals(2, third.counter.toLong())

        assertEquals(initialDtos[0], fourth.movieDto)
        assertEquals(fourth.counter.toLong(), 1)

        assertEquals(initialDtos[1], fifth.movieDto)
        assertEquals(fifth.counter.toLong(), 1)
    }
}